File: src/Python/core/metrics.py
from __future__ import annotations
import logging
import time
from typing import Protocol, Optional

class RunScope(Protocol):
    def on_step_start(self, index: int, label: str) -> None: ...
    def on_step_end(self, index: int, label: str, elapsed_nanos: int, success: bool) -> None: ...
    def on_step_error(self, index: int, label: str, error: BaseException) -> None: ...
    def on_jump(self, from_label: str, to_label: str, delay_ms: int) -> None: ...
    def on_pipeline_end(self, success: bool, elapsed_nanos: int, error: Optional[BaseException]) -> None: ...

class Metrics(Protocol):
    def on_pipeline_start(self, pipeline_name: str, run_id: str, start_label: Optional[str]) -> RunScope: ...

class NoopMetrics:
    def on_pipeline_start(self, pipeline_name: str, run_id: str, start_label: Optional[str]) -> RunScope:
        class _Run:
            def on_step_start(self, i, l): pass
            def on_step_end(self, i, l, e, s): pass
            def on_step_error(self, i, l, err): pass
            def on_jump(self, f, t, d): pass
            def on_pipeline_end(self, success, nanos, err): pass
        return _Run()

class LoggingMetrics:
    """Simple out-of-the-box metrics using :mod:`logging`."""
    def __init__(self, logger: Optional[logging.Logger]=None, level: int=logging.INFO):
        self.log = logger or logging.getLogger("pipeline")
        self.level = level

    def on_pipeline_start(self, pipeline_name: str, run_id: str, start_label: Optional[str]) -> RunScope:
        log, level = self.log, self.level
        if log.isEnabledFor(level):
            log.log(level, f"pipeline.start name={pipeline_name} runId={run_id} startLabel={start_label}")

        class _Run:
            def on_step_start(self, index, label):
                if log.isEnabledFor(level):
                    log.log(level, f"step.start name={pipeline_name} runId={run_id} index={index} label={label}")
            def on_step_end(self, index, label, elapsed_nanos, success):
                if log.isEnabledFor(level):
                    log.log(level, f"step.end   name={pipeline_name} runId={run_id} index={index} label={label} durMs={elapsed_nanos/1_000_000:.3f} success={success}")
            def on_step_error(self, index, label, error):
                log.warning(f"step.error name={pipeline_name} runId={run_id} index={index} label={label}", exc_info=error)
            def on_jump(self, from_label, to_label, delay_ms):
                if log.isEnabledFor(level):
                    log.log(level, f"step.jump  name={pipeline_name} runId={run_id} from={from_label} to={to_label} delayMs={delay_ms}")
            def on_pipeline_end(self, success, nanos, error):
                if error is None:
                    if log.isEnabledFor(level):
                        log.log(level, f"pipeline.end name={pipeline_name} runId={run_id} durMs={nanos/1_000_000:.3f} success={success}")
                else:
                    log.warning(f"pipeline.end name={pipeline_name} runId={run_id} durMs={nanos/1_000_000:.3f} success={success}", exc_info=error)
        return _Run()

File: src/typescript/core/metrics.ts
export interface RunScope {
  onStepStart(index: number, label: string): void;
  onStepEnd(index: number, label: string, elapsedNanos: number, success: boolean): void;
  onStepError(index: number, label: string, error: unknown): void;
  onJump(fromLabel: string, toLabel: string, delayMillis: number): void;
  onPipelineEnd(success: boolean, elapsedNanos: number, error?: unknown): void;
}

export interface Metrics {
  onPipelineStart(pipelineName: string, runId: string, startLabel?: string | null): RunScope;
}

export class NoopMetrics implements Metrics {
  onPipelineStart(pipelineName: string, runId: string, startLabel?: string | null): RunScope {
    const scope: RunScope = {
      onStepStart() {},
      onStepEnd() {},
      onStepError() {},
      onJump() {},
      onPipelineEnd() {}
    };
    return scope;
  }
}

export type LogLevel = "debug" | "info" | "log" | "warn" | "error";

export class LoggingMetrics implements Metrics {
  private logger: Console;
  private level: LogLevel;

  constructor(logger: Console = console, level: LogLevel = "info") {
    this.logger = logger;
    this.level = level;
  }

  onPipelineStart(pipelineName: string, runId: string, startLabel?: string | null): RunScope {
    const log = this.logger;
    const level = this.level;
    if (typeof (log as any)[level] === "function") {
      (log as any)[level](`pipeline.start name=${pipelineName} runId=${runId} startLabel=${startLabel}`);
    }

    const scope: RunScope = {
      onStepStart: (index, label) => {
        (log as any)[level](`step.start name=${pipelineName} runId=${runId} index=${index} label=${label}`);
      },
      onStepEnd: (index, label, elapsedNanos, success) => {
        const elapsedMillis = (elapsedNanos / 1_000_000).toFixed(3);
        (log as any)[level](`step.end   name=${pipelineName} runId=${runId} index=${index} label=${label} durMs=${elapsedMillis} success=${success}`);
      },
      onStepError: (index, label, error) => {
        log.warn(`step.error name=${pipelineName} runId=${runId} index=${index} label=${label}`, error);
      },
      onJump: (fromLabel, toLabel, delayMillis) => {
        (log as any)[level](`step.jump  name=${pipelineName} runId=${runId} from=${fromLabel} to=${toLabel} delayMs=${delayMillis}`);
      },
      onPipelineEnd: (success, elapsedNanos, error) => {
        const elapsedMillis = (elapsedNanos / 1_000_000).toFixed(3);
        if (error == null) {
          (log as any)[level](`pipeline.end name=${pipelineName} runId=${runId} durMs=${elapsedMillis} success=${success}`);
        } else {
          log.warn(`pipeline.end name=${pipelineName} runId=${runId} durMs=${elapsedMillis} success=${success}`, error);
        }
      }
    };
    return scope;
  }
}

export function nowNanos(): number {
  if (typeof performance !== "undefined" && typeof performance.now === "function") {
    return Math.floor(performance.now() * 1_000_000);
  }
  const dateMillis = Date.now();
  return dateMillis * 1_000_000;
}
